### ISSUE TITLE
`filter!` corrupts `SparseVector`

### ISSUE THREAD

**Issue by CameronBieganek** [JuliaLang/julia#53920]
If you try to use `filter!` on a `SparseVector`, it fails with an internal method error for `deleteat!`. And then your sparse vector is corrupted:

```julia
julia> using SparseArrays

julia> x = SparseVector(5, collect(1:2:7), collect(1:4))
5-element SparseVector{Int64, Int64} with 4 stored entries:
  [1]  =  1
  [3]  =  2
  [5]  =  3
  [7]  =  4

julia> filter!(iseven, x)
ERROR: MethodError: no method matching deleteat!(::SparseVector{Int64, Int64}, ::UnitRange{Int64})

Closest candidates are:
  deleteat!(::Vector, ::AbstractUnitRange{<:Integer})
   @ Base array.jl:1525
  deleteat!(::Vector, ::AbstractVector)
   @ Base array.jl:1567
  deleteat!(::Vector, ::Any)
   @ Base array.jl:1566
  ...

Stacktrace:
 [1] filter!(f::typeof(iseven), a::SparseVector{Int64, Int64})
   @ Base .\array.jl:2662
 [2] top-level scope
   @ REPL[4]:1

julia> x
5-element SparseVector{Int64, Int64} with 6 stored entries:
  [1]  =  0
  [2]  =  2
  [3]  =  0
  [4]  =  3
  [5]  =  3
  [7]  =  4
```

**Comment by jakobnissen** [comment]
Is this really a bug? This issue can happen whenever you have a generic function that temporarily causes its arguments to be in an invalid state - because it's always a possibility that the later operations that restore the correct state to the argument throws an error. So there are only two possibilities, really:
1. Never write generic methods that can cause an argument to be in an invalid state temporarily
2. Accept that after an exception is thrown in a function, you can have an invalid state

The first option is quite extreme and has far-reaching implications.

**Comment by CameronBieganek** [comment]
> Is this really a bug? This issue can happen whenever you have a generic function that temporarily causes its arguments to be in an invalid state - because it's always a possibility that the later operations that restore the correct state to the argument throws an error.

To more specifically respond to your question, I would expect the following to work, but it does not:

```julia
julia> x = sparsevec(1:2:7, 1:4, 8)
8-element SparseVector{Int64, Int64} with 4 stored entries:
  [1]  =  1
  [3]  =  2
  [5]  =  3
  [7]  =  4

julia> x[3]
2

julia> try
           filter!(iseven, x)
       catch
       end

julia> x[3]
0
```

**Comment by jakobnissen** [comment]
Regarding `filter(f, ::SparseVector)`, I don't understand why you would expect it to give you a dense vector, or zero out some entries. That's not how `filter` works, and it's not how it's documented to work. It's documented to return a copy of the collection without elements for which `f` returns `false`, and that's exactly what it does. From what I can tell, `filter` behaves correctly.

Regarding the try-catch thing, that's not possible to have that work in general. For that to work, there would need to be some enforcement that instances of some abstract type implements some given methods. There has been some discussion about this, but it's not part of Julia yet.

**Comment by adienes** [comment]
> But perhaps that expectation is unreasonable. 🤷‍♂️ Maybe I just have to accept the fact that Julia is a procedural language with lots of side effects.

I mean

isn't that the whole point of distinguishing `filter` from `filter!` ? you can get the behavior you want by just
```
x2 = copy(x)
filter!(iseven, x2)
x = x2
```

**Comment by jakobnissen** [comment]
The indices of the values change, because the removed elements are removed, not set to zero. If you remove e.g. the element at index 2 from a vector, then what was previously at element 3 will move down to be in position 2. This is the same as an ordinary Vector.
A sparse vector is not like a Dict - it's rather an alternative implementation of something like Vector. If the indices were preserved, then no elements would have been removed, which would have been wrong according to the docstring.
In other words, the elements not printed when you display the sparse vector are not missing elements. They exist, but are zero. This is different from a Dict.

**Comment by adienes** [comment]
> That obviously doesn't work:

I stand corrected -- I should have tried it before posting

> but apparently I have not been admitted to the enlightened circle of Julia developers who understand the true meaning of Julia's vague docstrings.

I apologize for having implied any such thing

w.r.t. to the example at hand

```
julia> x = sparsevec(1:2:7, 1:4, 8)
8-element SparseVector{Int64, Int64} with 4 stored entries:
  [1]  =  1
  [3]  =  2
  [5]  =  3
  [7]  =  4

julia> filter(iseven, x)
6-element SparseVector{Int64, Int64} with 2 stored entries:
  [2]  =  2
  [5]  =  4
```

I agree this is pretty confusing, and I had to do a double take. although after staring at it a bit it does seem correct. what was tripping me up was forgetting that of course in this case the `iseven` predicate _includes the zeros_. The behavior seems a little bit more clear when I make the predicate exclude zeros

```
julia> filter(v -> iseven(v) && !iszero(v), x)
2-element SparseVector{Int64, Int64} with 2 stored entries:
  [1]  =  2
  [2]  =  4
```

**Comment by mbauman** [comment]
I agree this is a bug — it's not just sparse, this method will corrupt _all_ vectors that are mutable but not mutably-sized.

There are three classes of functions:
* Primitives and some other methods are only defined on concrete types: `deleteat!`, `insert!`, `push!`, `pushfirst`, `pop!`, `popat!`, `popfirst!`, `empty!`, `splice!`
* Some abstract methods are implemented for `AbstractVector` but would immediately fail (because they effectively `resize!` first): `append!`, `prepend!`
* But then there are some abstract methods that will fail halfway through: `filter!`, `intersect!`, `union!`, `symdiff!`

I may have missed a few, but this isn't an abstract "yeah, anything can fail in the middle" sort of thing.  We **know** that many `AbstractVectors` are mutable but not mutable-length.  I think all these methods should check for applicability of what they expect to use to `resize!` before doing any preparatory work.
